src/python/zquantum/core/wavefunction.py

```diff
@@ -0,0 +1,142 @@
+from math import log2
+from typing import Any, Dict, List, Union
+from warnings import warn
+
+import numpy as np
+from sympy import Abs, Matrix, Symbol
+
+
+def _cast_sympy_matrix_to_numpy(sympy_matrix, complex=False):
+    new_type = np.complex128 if complex else np.float64
+
+    try:
+        return np.array(sympy_matrix, dtype=new_type).flatten()
+    except TypeError:
+        return np.array(sympy_matrix, dtype=object).flatten()
+
+
+class Wavefunction:
```

Comment:
Overall I'm afraid that insisting on using sympy wavefunctions might kill the performance.
I don't have a good solution for that, but I think it's worth having a discussion about it.
I mean – in general in the whole codebase, not only the wavefunctions themselves.
CC: @dexter2206 @alexjuda

Reply:
I thought about this too, but didn't want to fall for optimization without running any benchmarks. Do you feel like we should prepare some experiments and measure the performance before going with this refactor?

Reply:
my only argument for using sympy is the ability to calculate the final output of the circuit only once and never again, since we can depend on Symbol substitutions instead of repeating matrix multiplications for every run. This will be specially useful when running deep circuits.

Reply:
@alexjuda – I don't think it should be a blocker and also I feel that it's a more general problem with other places in our codebase as well. Will create a ticket for that.

@Shiro-Raven I haven't thought about it, but that could be actually very efficient!

Reply:
@mstechly @alexjuda We decided with @Shiro-Raven  to use similar approach as when converting circuits to unitaries. In current implementation we use numpy objects if possible and only fallback to sympy if there are any free  symbols presents in wavefunction.

The decision was made to make sure that the interface and inner workings of our wavefunction are precisely the same  as the PyQuil's wavefunction.